ODH release automation

AjayJagan opened the pull request:
A set of github actions/ scripts to automate the entire release process in github.

## Description
This change enables us to automate the odh release process through a set of github actions and scripts.
A detailed explanation can be found [here](https://docs.google.com/document/d/1cmMNGCKU9r4HUPtR16g-dqHL7UqPBDIpE72pXRWb-FA/edit).

AjayJagan commented:
The idea here is to create a draft PR so that we can start the review process and also find any misfits in the code.

AjayJagan commented on .github/workflows/release-gh-publish.yaml:
```diff
@@ -0,0 +1,88 @@
+        uses: peter-evans/create-pull-request@v6
+        with:
+          path: ./community-operators-prod
+          token: <PAT> # We need a token with repo rights
```
Here, for some reason the token generated by github app does not work. I created an issue in the library: https://github.com/peter-evans/create-pull-request/issues/2848 .
Also I created a shell script to mimic the pr creation process, but unfortunately even that flow works with a PAT and not with a token generated from GH app. Any suggestions/help here is much appreciated.

AjayJagan replied:
the PAT to be used needs the `repo` scope.

bartoszmajsak commented:
Happy to review it once I'm back in a week. Assuming it can wait.

VaishnaviHire commented:
@AjayJagan Feel free to also add example tracker [here](https://github.com/opendatahub-io/opendatahub-community/issues) for testing

AjayJagan commented:
> @AjayJagan Feel free to also add example tracker [here](https://github.com/opendatahub-io/opendatahub-community/issues) for testing

https://github.com/AjayJagan/notes-operator/issues/2

AjayJagan commented on .github/scripts/wait-for-checks.sh:
```diff
@@ -0,0 +1,25 @@
+
+while $(gh pr checks "$1" | grep -q -v 'tide' | grep -q 'pending'); do
+  printf ":stopwatch: PR checks still pending, retrying in 10 seconds...\n"
+  sleep 10
```
true. We could increase the interval to 10 minutes once or so.
I checked the `gh pr checks` for https://github.com/opendatahub-io/opendatahub-operator/pull/991
and here is what I get
for `gh pr checks 991 -R opendatahub-io/opendatahub-operator | grep pending` I get
```tide	pending	0	https://prow.ci.openshift.org/pr?query=is%3Apr+repo%3Aopendatahub-io%2Fopendatahub-operator+author%3Azdtsw+head%3Ajira%2F443_3	Not mergeable. Needs approved, lgtm labels.```
so I though we could do an inverse of tide(which will remove tide from the checks) and grep for pending... if present  we could sleep and repeat again

AjayJagan replied:
checking pr #992 which is currently running checks
for `gh pr checks 992 -R opendatahub-io/opendatahub-operator | grep -v tide | grep pending`

I get

```ci/prow/ci-index	pending	0	https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_opendatahub-operator/992/pull-ci-opendatahub-io-opendatahub-operator-incubation-ci-index/1785239301668212736	Job triggered.                     BaseSHA:ec07da31c18b26885daa006c964ae3ea450a0e6f
ci/prow/opendatahub-operator-e2e	pending	0	https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_opendatahub-operator/992/pull-ci-opendatahub-io-opendatahub-operator-incubation-opendatahub-operator-e2e/1785239301773070336	Job triggered.                     BaseSHA:ec07da31c18b26885daa006c964ae3ea450a0e6f```

AjayJagan replied:
the other way we could handle this is, manually wait for the checks to complete and a pr close with a particular comment should trigger it?

AjayJagan commented on .github/workflows/release-branch.yaml:
```diff
@@ -0,0 +1,31 @@
+name: "Release: Create release branch"
```
yea maybe we could use a different naming 😅 
But the idea is after creating a new branch with the changes, we wait for every checks to finish manually. So I did this because, if in case there is a failure, then we can stop at this point and prevent a release in github

AjayJagan commented on .github/workflows/release-gh-publish.yaml:
```diff
@@ -0,0 +1,88 @@
+          pat-token: ${{ steps.generate-token.outputs.token }}
+      - name: Create and push version tags
+        run: |
+          git config --global user.email 41898282+github-actions[bot]@users.noreply.github.com
```
checking now

AjayJagan replied:
so after taking a look. here we only tag right? Or do we create labels as well. AFAIK there is no gh cli way to create tags. Let me know if I am wrong here.

AjayJagan commented on .github/workflows/release-test-e2e.yaml:
```diff
@@ -0,0 +1,58 @@
+      - name: Create test release pr
+        uses: ./.github/actions/create-release-pr
+        with:
+          pr-branch: "odh-release/e2e-test"
```
so the `create-release-pr` is a reusable action used in 2 places. That is why it passes the pr-branch name.

AjayJagan commented on Makefile:
```diff
@@ -236,7 +236,7 @@ KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/k
 kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
 $(KUSTOMIZE): $(LOCALBIN)
-	test -s $(KUSTOMIZE) || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | sh -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
+	test -s $(KUSTOMIZE) || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
```
hmm now I dont remember why it broke and I had to change it. Let me check this once

AjayJagan commented on .github/workflows/release-test-e2e.yaml:
```diff
@@ -0,0 +1,58 @@
+          delete-branch: true
+      - name: Push version tag to quay.io
+        run: |
+          podman login -u <username> -p ${{ secrets.PODMAN_PASSWORD }} quay.io/opendatahub # Store PODMAN_PASSWORD of opendatahub-io here as a secret.
+          podman pull quay.io/opendatahub/opendatahub-operator:pr-${{ env.PULL_REQUEST_NUMBER }}
+          podman push quay.io/opendatahub/opendatahub-operator:pr-${{ env.PULL_REQUEST_NUMBER }} quay.io/opendatahub/opendatahub-operator:${{ env.VERSION }}
+          echo "Successfully push tags to podman with version: ${{ env.VERSION }}"
```
great! testing this now

AjayJagan replied:
alternatively we could use `skopeo` ?

AjayJagan commented on .github/workflows/release-update-versions.yaml:
```diff
@@ -0,0 +1,46 @@
+          sed -i -e "s|createdAt.*|createdAt: \"$(date +"%Y-%-m-%dT00:00:00Z")\"|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|name: opendatahub-operator.v.*|name: opendatahub-operator.v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|version: $CURRENT_VERSION.*|version: $NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|replaces.*|replaces: opendatahub-operator.v$CURRENT_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
```
will remove it 👍

AjayJagan commented on .github/workflows/release-update-versions.yaml:
```diff
@@ -0,0 +1,46 @@
+          sed -i -e "s|name: opendatahub-operator.v.*|name: opendatahub-operator.v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|version: $CURRENT_VERSION.*|version: $NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|replaces.*|replaces: opendatahub-operator.v$CURRENT_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|olm.skipRange:.*|olm.skipRange: \'>=$CURRENT_VERSION <$NEW_VERSION\'|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
```
agreed 👍

AjayJagan commented on Makefile:
```diff
@@ -236,7 +236,7 @@ KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/k
 kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
 $(KUSTOMIZE): $(LOCALBIN)
-	test -s $(KUSTOMIZE) || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | sh -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
+	test -s $(KUSTOMIZE) || { curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN); }
```
yes that is the reason

AjayJagan replied:
the error is 
```test -s /home/runner/work/opendatahub-operator/opendatahub-operator/bin/kustomize || { curl -s "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh" | sh -s -- 3.8.7 /home/runner/work/opendatahub-operator/opendatahub-operator/bin; }
sh: 33: Syntax error: "(" unexpected (expecting "then")```

ykaliuta replied:
That is actually ok. The script has bash's shebang so using sh to execute it is a mistake. Debians by default use dash as the system shell which is pretty much posix

zdtsw commented:
before going to update all the small detail implementation comments, I would like to have an overall understanding for the solution.
several parts bother from yesterday's review: 
- the only reason I see we have multiple workflow is 1)we need a dry-run PR to check the changes with e2e tests 2)we need the formal PR with release branch and updated new version. thus, i do not know why we have split the 2nd one into 3 flows. if the plan is one flow should only start after a successful run of the other flow, cannot this be done in side of the same one but as different jobs? isnt that much easier to pass env variable or even use steps' outputs?
- another one is, this shared-env part.  I still would like to know why exactly chose to use do it this way. 
if the dry-run workflow has inputs of version + trackerURL, cannot these be added into the PR or commit message? since you are already checking if the content of such in order to "start job" or not, these should be easier to get rather than manipulate(get mostly) actions variables

AjayJagan commented:
> before going to update all the small detail implementation comments, I would like to have an overall understanding for the solution. several parts bother from yesterday's review:
> 
> * the only reason I see we have multiple workflow is 1)we need a dry-run PR to check the changes with e2e tests 2)we need the formal PR with release branch and updated new version. thus, i do not know why we have split the 2nd one into 3 flows. if the plan is one flow should only start after a successful run of the other flow, cannot this be done in side of the same one but as different jobs? isnt that much easier to pass env variable or even use steps' outputs?
> * another one is, this shared-env part.  I still would like to know why exactly chose to use do it this way.
>   if the dry-run workflow has inputs of version + trackerURL, cannot these be added into the PR or commit message? since you are already checking if the content of such in order to "start job" or not, these should be easier to get rather than manipulate(get mostly) actions variables

So if I understand this correct, there can be 1 workflow with 3 jobs.
job 1 -> will do the dry run(if failed exit immediately)
job 2 -> will create the real pr for updating the manifests
job 3 -> will create the version updates
hmm this looks better and yea it removes the headache of sharing the env vars

I did not think in this angle https://github.com/opendatahub-io/opendatahub-operator/pull/988#discussion_r1583302953. So I think shared env is not the way to do it. I can create a comment with a particular message and read from it. That would make things much easier.

Considering the above changes, can I start the work? let me know if this is okay 
And thanks for the suggestions.

AjayJagan commented on .github/workflows/release-update-versions.yaml:
```diff
@@ -0,0 +1,46 @@
+          sed -i -e "s/^VERSION ?=.*/VERSION ?= $NEW_VERSION/g" Makefile
+          sed -i -e "s|containerImage.*|containerImage: quay.io/opendatahub/opendatahub-operator:v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|createdAt.*|createdAt: \"$(date +"%Y-%-m-%dT00:00:00Z")\"|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|name: opendatahub-operator.v.*|name: opendatahub-operator.v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|version: $CURRENT_VERSION.*|version: $NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
```
you mean if I run `make generate manifests` it should get generated?

AjayJagan replied:
curious question -> It was my understanding that the csv in `config/manifests` has to be manually updated and then when running a `make bundle` it applies these changes in the bundle/manifests. Or am I missing any commands 🤔

AjayJagan commented:
Hey @zdtsw , I have changed to workflows and reduced them to 2 workflows with multiple jobs. Also I have made some changes in the code. Let me know if it looks good or it can be made better. Thanks :)

ykaliuta commented on .github/scripts/wait-for-checks.sh:
```diff
@@ -0,0 +1,25 @@
+
+sleep 10
+
+while [ $(gh pr checks "$1" | grep -v 'tide' | grep 'pending' | wc -m) -gt 0 ]; do
```
is it
```
while gh pr checks "$1" | grep -v 'tide' | grep -q 'pending'; do ...
```
?

ykaliuta commented on .github/scripts/wait-for-checks.sh:
```diff
@@ -0,0 +1,25 @@
+  sleep 30 # replace with 60000
+done
+
+if [ $(gh pr checks "$1" | grep 'fail' | wc -m) -gt 0 ]; then
```
ditto.

But it looks that you are interested in the second column only, you can grep some other parts by accident.
Would the wrapper work? 
```
# would be shorter if negated, just exit 1 in if
# choose better name
status_exists() {
    gh pr checks $1 | awk "{if (\$2 == \"$2\") found=1} END{if (!found) exit 1}"
}

# probably you do not expect PR number with spaces?
status_exists $1 fail && { echo "check failed"; exit 1}
```

ykaliuta commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,90 @@
+          body-includes: "#Release#"
+      - name: Set version/tracker-url in env
+        run: |
+          echo "VERSION=$(echo "${{ steps.fc.outputs.comment-body }}" | grep version | cut -d "=" -f2)" >> $GITHUB_ENV
```
something like ```awk -F= '$1 == "version" {print $2}'``` ?

AjayJagan replied:
this works 
`awk -F' ' '{print $2}' | awk -F= '$1 == "version"{print $2}'`

AjayJagan commented on .github/workflows/pre-release.yaml:
```diff
@@ -0,0 +1,108 @@
+          delete-branch: true
+      - name: Push version tag to quay.io
+        run: |
+          skopeo login -u <QUAY_USERNAME> -p ${{ secrets.QUAY_PASSWORD }} quay.io/opendatahub # Store PODMAN_PASSWORD of opendatahub-io here as a secret.
```
need to replace with username and password for opendatahub account

AjayJagan commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,90 @@
+        uses: peter-evans/create-pull-request@v6
+        with:
+          path: ./community-operators-prod
+          token: <PAT> # We need a token with repo rights
```
The token generated by gh-app is not working in this case. So thinking of using a PAT with repo rights(minimal)

AjayJagan commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,90 @@
+      - name: Checkout redhat-openshift-ecosystem/community-operators-prod
+        uses: actions/checkout@v4
+        with:
+          token: <PAT> # We need a token with repo rights
```
The token generated by gh-app is not working in this case. So thinking of using a PAT with repo rights(minimal)

AjayJagan commented:
/retest

ykaliuta commented on .github/scripts/wait-for-checks.sh:
```diff
@@ -2,24 +2,25 @@
+done
 
-printf "!!An unknown error occurred!!\n"
+pr_has_status $1 fail && { echo "!!PR checks failed!!"; exit 1; }
```
I'm wondering, since ```gh pr checks``` has exist status 0 (success) when there are no failed tests and non-0 otherwise,  if it's sufficient here to just run it without extra logic? Probably, "unknown error" or "failed checks" will be clear from the log?

VaishnaviHire commented:
I think we should not block this for operatorhub PR automation. 
Lets have the automation setup for ODH release and have a follow-up PR to update operatorhub changes.

AjayJagan commented:
> I think we should not block this for operatorhub PR automation. Lets have the automation setup for ODH release and have a follow-up PR to update operatorhub changes.

@VaishnaviHire, with the secrets set for quay and gh app, we should be good to test this :)

zdtsw commented on .github/scripts/validate-semver.sh:
```diff
@@ -0,0 +1,23 @@
+
+sem_ver_pattern="^[vV](0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(\\-[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?(\\+[0-9A-Za-z-]+(\\.[0-9A-Za-z-]+)*)?$"
+
+die () {
+    echo >&2 "$@"
+    exit 1
+}
```
why we have mixed usage of die here and exit 1 other places?

AjayJagan replied:
so I had reused the script used in istio-workspaces: https://github.com/maistra/istio-workspace/blob/master/scripts/release/validate.sh. I can use a simple echo + exit 1 in the validate-semver.sh if that's okay.

zdtsw commented on .github/workflows/pre-release.yaml:
```diff
@@ -0,0 +1,108 @@
+          pull-request-number: ${{ steps.cpr-dry-run.outputs.pull-request-number }}
+          comment: Auto-closing pull request after success checks
+          delete-branch: true
+      - name: Push version tag to quay.io
+        run: |
+          skopeo login -u ${{ secrets.QUAY_ID }} -p ${{ secrets.QUAY_TOKEN }} quay.io/{{ secrets.QUAY_ORG }}
+          skopeo copy docker://quay.io/{{ secrets.QUAY_ORG }}/opendatahub-operator:pr-${{ steps.cpr-dry-run.outputs.pull-request-number }} docker://quay.io/{{ secrets.QUAY_ORG }}/opendatahub-operator:${{ env.VERSION }}
+          echo "Successfully updated tag to quay.io with version: ${{ env.VERSION }}"
```
https://github.com/marketplace/actions/skopeo-copy-action ?

AjayJagan replied:
I saw that action, but looked like a very unsupported action 😅  so I did not use it. Do you think it is okay to use?

zdtsw replied:
ok, keep your part, i am fine with this.

AjayJagan replied:
looking at the code, https://github.com/coscene-io/skopeo-copy-action/blob/main/action.yml
they do pretty much what we do manually

zdtsw commented on .github/workflows/pre-release.yaml:
```diff
@@ -0,0 +1,108 @@
+          git push -f origin odh-${{ env.VERSION }}
+      - uses: ./.github/actions/update-manifest-branches
+      - name: Update versions in relevant files
+        run: |
+          NEW_VERSION=${{ env.VERSION }}
+          CURRENT_VERSION=$(cat Makefile | grep -w "VERSION ?=" | cut -d ' ' -f 3)
+          sed -i -e "s/^VERSION ?=.*/VERSION ?= $NEW_VERSION/g" Makefile
+          sed -i -e "s|containerImage.*|containerImage: quay.io/opendatahub/opendatahub-operator:v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|createdAt.*|createdAt: \"$(date +"%Y-%-m-%dT00:00:00Z")\"|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|name: opendatahub-operator.v.*|name: opendatahub-operator.v$NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|version: $CURRENT_VERSION.*|version: $NEW_VERSION|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
+          sed -i -e "s|olm.skipRange:.*|olm.skipRange: \'>=1.0.0 <$NEW_VERSION\'|g" config/manifests/bases/opendatahub-operator.clusterserviceversion.yaml
```
can we move these into on scripts?

AjayJagan replied:
will do

VaishnaviHire reviewed (approved):
/lgtm
